src/renderers/opengl/mod.rs

```diff
@@ -89,23 +121,86 @@ impl GlCache {
     }
 }
 
-type ErasedNodeRenderer = Box<dyn Fn(&OpenglRenderer, &dyn Node)>;
+pub type OpenglRenderer = ExtOpenglRenderer<(), DefaultCustomRenderer>;
 
-pub struct OpenglRenderer {
+pub trait CustomRenderer {
+    type NodeData;
+    type Renderer;
+
+    fn render(
+        &self,
+        renderer: &Self::Renderer,
+        node: &ExtInoxNode<Self::NodeData>,
+        node_data: &Self::NodeData,
+    );
+}
+
+pub struct DefaultCustomRenderer;
+
+impl CustomRenderer for DefaultCustomRenderer {
+    type NodeData = ();
+    type Renderer = OpenglRenderer;
+
+    fn render(
+        &self,
+        _renderer: &Self::Renderer,
+        _node: &ExtInoxNode<Self::NodeData>,
+        _node_data: &Self::NodeData,
+    ) {
+    }
+}
+
+pub fn opengl_renderer(
+    gl: glow::Context,
+    nodes: InoxNodeTree,
+    textures: Vec<ModelTexture>,
+) -> OpenglRenderer {
+    ExtOpenglRenderer::new(gl, nodes, textures, DefaultCustomRenderer)
+}
+
+pub fn opengl_renderer_ext<T, R>(
+    gl: glow::Context,
+    nodes: ExtInoxNodeTree<T>,
+    textures: Vec<ModelTexture>,
+    custom_renderer: R,
+) -> ExtOpenglRenderer<T, R>
+where
+    R: CustomRenderer<NodeData = T, Renderer = ExtOpenglRenderer<T, R>>,
+{
+    ExtOpenglRenderer::new(gl, nodes, textures, custom_renderer)
+}
+
+pub struct ExtOpenglRenderer<T, R>
```

Comment:
Would it make much more sense to have the `opengl_renderer` function as `ExtOpenglRenderer::default_renderer()` or `ExtOpenglRenderer::new()` and `opengl_renderer_ext` as `OpenglRenderer::custom_renderer()`?

Reply:
The first time is a bit confusing and verbose. `Ext` feels unnecessary for a renderer that only supports built-in nodes.

Definitely not for the second option since `OpenglRenderer` is a type alias.

I think this is fine. I'm just doing the same thing that `glam` does with helper functions like `vec2` and `vec3` which are just wrappers for `Vec2::new()` and `Vec3::new()`.

Reply:
> I think this is fine. I'm just doing the same thing that `glam` does with helper functions like `vec2` and `vec3` which are just wrappers for `Vec2::new()` and `Vec3::new()`.

To be fair, kinda mixed feelings. It just feels like when you have a long function name it's much more intended to do certain action other than constructing a struct? `vec2` and `vec3` are short names so they're not big problems.

Reply:
When we have `OpenglRenderer::new()` users can just assume "Oh, it's to create a new struct!"

Reply:
That's not my point of view. Take a look at [xilem](https://github.com/linebender/xilem), it's filled with functions that are nothing but constructor wrappers. I think it takes part in what makes their architecture concise.

We don't need to do it, but in this case it makes things especially simple without much effort. :)

Reply:
> When we have `OpenglRenderer::new()` users can just assume "Oh, it's to create a new struct!"

In that regard, having `new` might even be misleading. `OpenglRenderer::new()` doesn't only initialize a struct, but also makes several OpenGL calls, meaning it would probably fail the program if the OpenGL context is not created.

Reply:
Oh, yeah...

Reply:
> meaning it would probably fail the program if the OpenGL context is not created.
We could warn them though?

Reply:
https://docs.rs/glutin/latest/glutin/display/enum.Display.html#method.new

Glutin does this:
> # [Safety](https://docs.rs/glutin/latest/glutin/display/enum.Display.html#safety)
> The display must point to the valid platform display and be valid for the entire lifetime of all Objects created with that display.
> The preference must contain pointers to the valid values if GLX or WGL specific options were used.

Reply:
True